lib/spack/spack/cmd/mirror.py

```diff
@@ -231,142 +234,225 @@ def mirror_list(args):
     mirrors.display()
 
 
-def _read_specs_from_file(filename):
-    specs = []
-    with open(filename, "r") as stream:
-        for i, string in enumerate(stream):
-            try:
-                s = Spec(string)
-                spack.repo.path.get_pkg_class(s.name)
-                specs.append(s)
-            except SpackError as e:
-                tty.debug(e)
-                tty.die("Parse error in %s, line %d:" % (filename, i + 1), ">>> " + string, str(e))
-    return specs
+def specs_from_text_file(filename, concretize=False):
+    """Return a list of specs read from a text file.
 
+    The file should contain one spec per line.
 
-def _determine_specs_to_mirror(args):
-    if args.specs and args.all:
-        raise SpackError(
-            "Cannot specify specs on command line if you" " chose to mirror all specs with '--all'"
-        )
-    elif args.file and args.all:
-        raise SpackError(
-            "Cannot specify specs with a file ('-f') if you"
-            " chose to mirror all specs with '--all'"
-        )
+    Args:
+        filename (str): name of the file containing the abstract specs.
+        concretize (bool): if True concretize the specs before returning
+            the list.
+    """
+    with open(filename, "r") as f:
+        specs_in_file = f.readlines()
+        specs_in_file = [s.strip() for s in specs_in_file]
+    return spack.cmd.parse_specs(" ".join(specs_in_file), concretize=concretize)
```

Comment:
Yes, it ensures that specs that are concretized are always treated in the same way, e.g. if they fail the concretization step they emit the same error with the same message. This, in the end, gives a more consistent behavior of the command.

Reply:
Got it, my original concern was that it would concretize all the specs together (which would not be desirable for mirrors). AFAIK our API doesn't make any guarantees one way or the other (e.g. I see no unit tests enforcing this or docstrings mentioning this behavior explicitly).